rust/src/error.rs

```diff
@@ -0,0 +1,396 @@
+            AdbcStatusCode::NotImplemented => write!(f, "Not Implemented"),
+            AdbcStatusCode::NotFound => write!(f, "Not Found"),
+            AdbcStatusCode::AlreadyExists => write!(f, "Already Exists"),
+            AdbcStatusCode::InvalidArguments => write!(f, "Invalid Arguments"),
+            AdbcStatusCode::InvalidState => write!(f, "Invalid State"),
+            AdbcStatusCode::InvalidData => write!(f, "Invalid Data"),
+            AdbcStatusCode::Integrity => write!(f, "Integrity"),
+            AdbcStatusCode::Internal => write!(f, "Internal Error"),
+            AdbcStatusCode::IO => write!(f, "IO Error"),
+            AdbcStatusCode::Cancelled => write!(f, "Cancelled"),
+            AdbcStatusCode::Timeout => write!(f, "Timeout"),
+            AdbcStatusCode::Unauthenticated => write!(f, "Unauthenticated"),
+            AdbcStatusCode::Unauthorized => write!(f, "Unauthorized"),
+        }
+    }
+}
+
+/// A detailed error message for an operation.
+#[repr(C)]
+#[derive(Debug, Copy, Clone)]
+pub struct FFI_AdbcError {
+    /// The error message.
+    pub message: *mut c_char,
+    /// A vendor-specific error code, if applicable.
+    pub vendor_code: i32,
+    /// A SQLSTATE error code, if provided, as defined by the
+    /// SQL:2003 standard.  If not set, it should be set to
+    /// "\\0\\0\\0\\0\\0".
+    pub sqlstate: [c_char; 5usize],
+    /// Release the contained error.
+    ///
+    /// Unlike other structures, this is an embedded callback to make it
+    /// easier for the driver manager and driver to cooperate.
+    pub release: Option<unsafe extern "C" fn(error: *mut Self)>,
+}
+
+impl FFI_AdbcError {
+    /// Create an empty error
+    pub fn empty() -> Self {
+        Self {
+            message: null_mut(),
+            vendor_code: -1,
+            sqlstate: ['\0' as c_char; 5],
+            release: None,
+        }
+    }
+
+    /// Create a new FFI_AdbcError.
+    ///
+    /// `vendor_code` defaults to -1 and `sql_state` defaults to zeros.
+    pub fn new(message: &str, vendor_code: Option<i32>, sqlstate: Option<[c_char; 5]>) -> Self {
+        Self {
+            message: CString::new(message).unwrap().into_raw(),
+            vendor_code: vendor_code.unwrap_or(-1),
+            sqlstate: sqlstate.unwrap_or(['\0' as c_char; 5]),
+            release: Some(drop_adbc_error),
+        }
+    }
+
+    /// Set an error message.
+    ///
+    /// # Safety
+    ///
+    /// If `dest` is null, no error is written. If `dest` is non-null, it must
+    /// be valid for writes.
+    pub unsafe fn set_message(dest: *mut Self, message: &str) {
+        if !dest.is_null() {
+            let error = Self::new(message, None, None);
+            unsafe { std::ptr::write_unaligned(dest, error) }
+        }
+    }
+}
+
+/// An error that can be converted into [FFI_AdbcError] and [AdbcStatusCode].
+///
+/// Can be used in combination with [check_err] when implementing ADBC FFI
+/// functions.
+pub trait AdbcError {
```

Comment:
Hmm. Instead of requiring boxing everywhere, what about having a concrete error type and just converting to/from the FFI version as necessary?

Also in general, I think it'd be better to compartmentalize the FFI bits the same way Go does, instead of assuming FFI as a fundamental part of the implementation

Reply:
I agree with this now. When I was thinking more FFI-centric this made a bit more sense. I have refactored to use a concrete struct in #478.